File: toolkit-cli/src/cbus_toolkit/calculator.py
"""Offline reproduction of C-Gate 3.4 CALCULATOR TEST's database calculation.

Catalogue metadata is supplied by the user, never bundled. The calculation is
for database units and deliberately preserves native edges: exact catalogue and
UnitType casing, last duplicate/alias wins, prefix-star fallback, Burden's literal
"0" comparison, the KEYGL5/SENTEMP4 burden exclusions, and ignoring impedance
below 10 ohms. It does not model cable resistance or hardware measurements.
"""
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import math
from pathlib import Path
from typing import Any, Iterable, Mapping
import xml.etree.ElementTree as ET

from .unitspec import _read_xml

MAX_UNITS = 10000


class CalculatorError(ValueError):
    """Invalid input or a network for which the native calculator cannot return results."""


@dataclass(frozen=True)
class CalculatorUnit:
    catalog_number: str | None
    unit_type: str
    burden: str | None = None
    switchable_supply_enabled: bool = False
    parameters: tuple[tuple[str, str], ...] = ()

    def __post_init__(self):
        if self.catalog_number is not None and not isinstance(self.catalog_number, str):
            raise CalculatorError("Catalogue number must be text or null")
        if not isinstance(self.unit_type, str) or not self.unit_type:
            raise CalculatorError("Unit type must be nonempty text")
        if self.burden is not None and not isinstance(self.burden, str):
            raise CalculatorError("Burden must be its native text value or null")
        if not isinstance(self.switchable_supply_enabled, bool):
            raise CalculatorError("Switchable supply enabled must be a boolean")
        try:
            parameters = tuple(self.parameters.items()) if isinstance(self.parameters, Mapping) else tuple(self.parameters)
        except TypeError as error:
            raise CalculatorError("Parameters must be ordered name/text-value pairs") from error
        if any(not isinstance(item, (tuple, list)) or len(item) != 2 or not all(isinstance(value, str) for value in item) for item in parameters):
            raise CalculatorError("Parameters must be ordered name/text-value pairs")
        if self.burden is not None and any(name in ("Burden", "HardwareBurdenMarker") for name, _value in parameters):
            raise CalculatorError("Supply burden once, either directly or in the ordered parameters")
        object.__setattr__(self, "parameters", tuple(tuple(item) for item in parameters))

    @classmethod
    def from_mapping(cls, value: Mapping[str, Any]) -> CalculatorUnit:
        unknown = set(value) - {"catalog_number", "unit_type", "burden", "switchable_supply_enabled", "parameters"}
        if unknown:
            raise CalculatorError("Unknown calculator unit fields: " + ", ".join(sorted(unknown)))
        try:
            return cls(**value)
        except TypeError as error:
            raise CalculatorError("Calculator units require catalogue number and unit type") from error

    def burden_enabled(self) -> bool:
        if self.unit_type.upper() in ("KEYGL5", "SENTEMP4"):
            return False
        if self.burden is not None:
            return self.burden != "0"
        for name, value in self.parameters:
            if name in ("Burden", "HardwareBurdenMarker"):
                return value != "0"
        return False


@dataclass(frozen=True)
class CalculatorLimits:
    min_impedance_ohms: int
    max_impedance_ohms: int
    max_supply_current_ma: int


@dataclass(frozen=True)
class CalculationResult:
    result: str
    current_supply_ma: int
    current_consumption_ma: int
    impedance_ohms: int
    exact_impedance_ohms: float
    units_calculated: int
    units_not_calculated: int
    limits: CalculatorLimits | None
    reasons: tuple[str, ...]
    catalog_sha256: str

    @property
    def passed(self) -> bool:
        return self.result == "OK"

    def native_values(self) -> dict[str, Any]:
        """The six fields returned by native CALCULATOR TEST, ready for comparison."""
        return {"result": self.result, "current_supply(mA)": self.current_supply_ma,
                "current_consumption(mA)": self.current_consumption_ma,
                "impedance(ohms)": float(self.impedance_ohms),
                "units_calculated": self.units_calculated,
                "units_not_calculated": self.units_not_calculated}


@dataclass(frozen=True)
class _ElectricalUnit:
    current_drawn: int
    current_supplied: int
    switchable_drawn: int
    switchable_supplied: int
    impedance: int
    switchable_declared: bool
    unit_types: tuple[str, ...]


def _number(element: ET.Element, field: str, default: int = 0) -> int:
    text = element.findtext(field)
    if text is None:
        return default
    try:
        value = int(text.strip(), 10)
    except (TypeError, ValueError) as error:
        raise CalculatorError(f"Invalid catalogue integer field {field}") from error
    if not -(1 << 31) <= value < (1 << 31):
        raise CalculatorError(f"Catalogue field {field} exceeds the native signed 32-bit range")
    return value


def _int32(value: int) -> int:
    return (value + (1 << 31)) % (1 << 32) - (1 << 31)


class CalculatorCatalog:
    """Native catalogue lookup and electrical totals, limited to 10,000 input units."""
    def __init__(self, entries: Mapping[str, _ElectricalUnit], limits: CalculatorLimits | None, sha256: str):
        self._entries = dict(entries)
        self.limits = limits
        self.sha256 = sha256
        # Native cW builds this set from the final alias/catalogue lookup map.
        self.switchable_unit_types = frozenset(unit_type for record in self._entries.values()
                                              if record.switchable_declared for unit_type in record.unit_types)

    @classmethod
    def load(cls, path: str | Path) -> CalculatorCatalog:
        path = Path(path)
        root = _read_xml(path)
        if root.tag != "CBusUnits":
            raise CalculatorError("Expected a CBusUnits catalogue")
        entries = {}
        def visit(unit: ET.Element):
            subunits = unit.find("SubUnits")
            if subunits is not None:
                for child in subunits.findall("Unit"):
                    visit(child)
                return
            record = _ElectricalUnit(
                _number(unit, "CurrentDrawn"), _number(unit, "CurrentSupplied"),
                _number(unit, "CurrentDrawnSwitchablePowerSupply"), _number(unit, "CurrentSuppliedSwitchablePowerSupply"),
                _number(unit, "Impedance"),
                unit.find("CurrentDrawnSwitchablePowerSupply") is not None and unit.find("CurrentSuppliedSwitchablePowerSupply") is not None,
                tuple(revision.findtext("UnitType", "") for revision in unit.findall("FirmwareRevisions/Revision")))
            catalog = unit.findtext("CatalogNumber", "")
            if catalog:
                entries[catalog] = record
            for alternative in unit.findtext("AlternativeCatalogNumbers", "").split(";"):
                if alternative:
                    entries[alternative] = record
        for unit in root.findall("Units/Unit"):
            visit(unit)
        calculator = root.find("Calculator")
        fields = ("MinImpedance", "MaxImpedance", "MaxSupplyCurrent")
        limits = CalculatorLimits(*(_number(calculator, name) for name in fields)) if calculator is not None and all(calculator.find(name) is not None for name in fields) else None
        return cls(entries, limits, hashlib.sha256(path.read_bytes()).hexdigest())

    def _lookup(self, catalog_number: str | None) -> _ElectricalUnit | None:
        if not catalog_number:
            return None
        record = self._entries.get(catalog_number)
        # Native only probes nonempty prefixes; a lone '*' is not a catch-all.
        while record is None and catalog_number:
            record = self._entries.get(catalog_number + "*")
            catalog_number = catalog_number[:-1]
        return record

    def calculate(self, units: Iterable[CalculatorUnit | Mapping[str, Any]]) -> CalculationResult:
        supply = consumption = known = unknown = 0
        conductance = 0.0
        for index, unit in enumerate(units):
            if index >= MAX_UNITS:
                raise CalculatorError(f"Calculator input exceeds the {MAX_UNITS}-unit limit")
            if isinstance(unit, Mapping):
                unit = CalculatorUnit.from_mapping(unit)
            if not isinstance(unit, CalculatorUnit):
                raise CalculatorError("Each calculator input must be a CalculatorUnit or mapping")
            record = self._lookup(unit.catalog_number)
            if record is None:
                unknown += 1
                continue
            if unit.unit_type in self.switchable_unit_types:
                if unit.switchable_supply_enabled:
                    supply = _int32(supply + record.switchable_supplied)
                else:
                    consumption = _int32(consumption + record.switchable_drawn)
            else:
                consumption = _int32(consumption + record.current_drawn)
                supply = _int32(supply + record.current_supplied)
            if record.impedance >= 10:
                conductance += 1.0 / record.impedance
            if unit.burden_enabled():
                conductance += 0.001
            known += 1
        if conductance == 0.0:
            error = CalculatorError("The field Impedance Accumulator cannot be zero.")
            error.units_calculated, error.units_not_calculated = known, unknown
            raise error
        impedance = 1.0 / conductance if conductance > 0 else 0.0
        reasons = []
        if self.limits is None:
            reasons.append("Catalogue calculator limits are missing or incomplete")
        else:
            if consumption > supply:
                reasons.append("Current consumption exceeds supply")
            if supply > self.limits.max_supply_current_ma:
                reasons.append("Supply current exceeds the catalogue maximum")
            if impedance < self.limits.min_impedance_ohms:
                reasons.append("Impedance is below the catalogue minimum")
            if impedance > self.limits.max_impedance_ohms:
                reasons.append("Impedance exceeds the catalogue maximum")
        if unknown:
            reasons.append("Some catalogue numbers were not found")
        return CalculationResult("FAILED" if reasons else "OK", supply, consumption,
                                 math.floor(impedance + 0.5), impedance, known, unknown,
                                 self.limits, tuple(reasons), self.sha256)


def calculate(catalog: str | Path, units: Iterable[CalculatorUnit | Mapping[str, Any]]) -> CalculationResult:
    return CalculatorCatalog.load(catalog).calculate(units)

File: rust/cbus-cgate/src/service/calculator.rs
//! Native-compatible database network electrical calculator.
//!
//! The arithmetic mirrors the independently accepted Python Toolkit
//! implementation and consumes an operator-supplied `cbusunits.xml`. It uses
//! database records only; no PCI traffic or physical measurement is involved.

use std::{collections::HashMap, path::Path};

use crate::{err, unitspec, Response, Server, Unit};

const MAX_UNITS: usize = 10_000;

#[derive(Debug, Clone)]
struct ElectricalUnit {
    current_drawn: i32,
    current_supplied: i32,
    switchable_drawn: i32,
    switchable_supplied: i32,
    impedance: i32,
    switchable_declared: bool,
    unit_types: Vec<String>,
}

#[derive(Debug, Clone, Copy)]
struct Limits {
    min_impedance: i32,
    max_impedance: i32,
    max_supply_current: i32,
}

struct Catalog {
    entries: HashMap<String, ElectricalUnit>,
    switchable_unit_types: std::collections::HashSet<String>,
    limits: Option<Limits>,
}

#[derive(Debug, Clone, Copy)]
struct ResultValues {
    passed: bool,
    supply: i32,
    consumption: i32,
    impedance: i64,
    known: usize,
    unknown: usize,
}

fn child<'a, 'input>(
    node: roxmltree::Node<'a, 'input>,
    name: &str,
) -> Option<roxmltree::Node<'a, 'input>> {
    node.children()
        .find(|child| child.is_element() && child.tag_name().name() == name)
}

fn text(node: roxmltree::Node<'_, '_>, name: &str) -> String {
    child(node, name)
        .and_then(|child| child.text())
        .unwrap_or("")
        .to_string()
}

fn number(node: roxmltree::Node<'_, '_>, name: &str) -> Result<i32, String> {
    let Some(value) = child(node, name).and_then(|child| child.text()) else {
        return Ok(0);
    };
    value
        .trim()
        .parse::<i32>()
        .map_err(|_| format!("Invalid catalogue integer field {name}"))
}

fn add_unit(
    node: roxmltree::Node<'_, '_>,
    entries: &mut HashMap<String, ElectricalUnit>,
) -> Result<(), String> {
    if let Some(subunits) = child(node, "SubUnits") {
        for unit in subunits
            .children()
            .filter(|node| node.is_element() && node.tag_name().name() == "Unit")
        {
            add_unit(unit, entries)?;
        }
        return Ok(());
    }
    let switchable_drawn = child(node, "CurrentDrawnSwitchablePowerSupply").is_some();
    let switchable_supplied = child(node, "CurrentSuppliedSwitchablePowerSupply").is_some();
    let unit_types = child(node, "FirmwareRevisions")
        .into_iter()
        .flat_map(|revisions| {
            revisions
                .children()
                .filter(|node| node.is_element() && node.tag_name().name() == "Revision")
        })
        .map(|revision| text(revision, "UnitType"))
        .collect::<Vec<_>>();
    let record = ElectricalUnit {
        current_drawn: number(node, "CurrentDrawn")?,
        current_supplied: number(node, "CurrentSupplied")?,
        switchable_drawn: number(node, "CurrentDrawnSwitchablePowerSupply")?,
        switchable_supplied: number(node, "CurrentSuppliedSwitchablePowerSupply")?,
        impedance: number(node, "Impedance")?,
        switchable_declared: switchable_drawn && switchable_supplied,
        unit_types,
    };
    let catalog_number = text(node, "CatalogNumber");
    if !catalog_number.is_empty() {
        entries.insert(catalog_number, record.clone());
    }
    for alternative in text(node, "AlternativeCatalogNumbers").split(';') {
        if !alternative.is_empty() {
            entries.insert(alternative.to_string(), record.clone());
        }
    }
    Ok(())
}

fn load_catalog(directory: &Path) -> Result<Catalog, String> {
    let xml = unitspec::read_xml_file(directory, "cbusunits.xml")?;
    let document = roxmltree::Document::parse(&xml)
        .map_err(|error| format!("Malformed calculator catalogue: {error}"))?;
    let root = document.root_element();
    if root.tag_name().name() != "CBusUnits" {
        return Err("cbusunits.xml is not a CBusUnits catalogue".to_string());
    }
    let mut entries = HashMap::new();
    if let Some(units) = child(root, "Units") {
        for unit in units
            .children()
            .filter(|node| node.is_element() && node.tag_name().name() == "Unit")
        {
            add_unit(unit, &mut entries)?;
        }
    }
    let switchable_unit_types = entries
        .values()
        .filter(|record| record.switchable_declared)
        .flat_map(|record| record.unit_types.iter().cloned())
        .collect();
    let limits = match child(root, "Calculator") {
        Some(calculator)
            if ["MinImpedance", "MaxImpedance", "MaxSupplyCurrent"]
                .iter()
                .all(|name| child(calculator, name).is_some()) =>
        {
            Some(Limits {
                min_impedance: number(calculator, "MinImpedance")?,
                max_impedance: number(calculator, "MaxImpedance")?,
                max_supply_current: number(calculator, "MaxSupplyCurrent")?,
            })
        }
        _ => None,
    };
    Ok(Catalog {
        entries,
        switchable_unit_types,
        limits,
    })
}

impl Catalog {
    fn lookup(&self, catalog_number: &str) -> Option<&ElectricalUnit> {
        if catalog_number.is_empty() {
            return None;
        }
        if let Some(record) = self.entries.get(catalog_number) {
            return Some(record);
        }
        let mut prefix = catalog_number.to_string();
        while !prefix.is_empty() {
            if let Some(record) = self.entries.get(&format!("{prefix}*")) {
                return Some(record);
            }
            prefix.pop();
        }
        None
    }

    fn calculate<'a>(&self, units: impl IntoIterator<Item = &'a Unit>) -> Option<ResultValues> {
        let mut supply = 0i32;
        let mut consumption = 0i32;
        let mut known = 0usize;
        let mut unknown = 0usize;
        let mut conductance = 0.0f64;
        for (index, unit) in units.into_iter().enumerate() {
            if index >= MAX_UNITS {
                return None;
            }
            let catalog_number = unit
                .fields
                .get("CatalogNumber")
                .map(String::as_str)
                .unwrap_or("");
            let Some(record) = self.lookup(catalog_number) else {
                unknown += 1;
                continue;
            };
            let unit_type = unit
                .fields
                .get("UnitType")
                .filter(|value| !value.is_empty())
                .map(String::as_str)
                .unwrap_or(&unit.unit_type);
            if self.switchable_unit_types.contains(unit_type) {
                let enabled = unit
                    .fields
                    .get("SwitchablePowerSupplyEnabled")
                    .is_some_and(|value| value.eq_ignore_ascii_case("true"));
                if enabled {
                    supply = supply.wrapping_add(record.switchable_supplied);
                } else {
                    consumption = consumption.wrapping_add(record.switchable_drawn);
                }
            } else {
                consumption = consumption.wrapping_add(record.current_drawn);
                supply = supply.wrapping_add(record.current_supplied);
            }
            if record.impedance >= 10 {
                conductance += 1.0 / f64::from(record.impedance);
            }
            let burden = unit
                .fields
                .get("Burden")
                .or_else(|| unit.fields.get("HardwareBurdenMarker"));
            if !matches!(
                unit_type.to_ascii_uppercase().as_str(),
                "KEYGL5" | "SENTEMP4"
            ) && burden.is_some_and(|value| value != "0")
            {
                conductance += 0.001;
            }
            known += 1;
        }
        if conductance == 0.0 {
            return None;
        }
        let exact_impedance = 1.0 / conductance;
        let impedance = (exact_impedance + 0.5).floor() as i64;
        let passed = self.limits.is_some_and(|limits| {
            consumption <= supply
                && supply <= limits.max_supply_current
                && exact_impedance >= f64::from(limits.min_impedance)
                && exact_impedance <= f64::from(limits.max_impedance)
                && unknown == 0
        });
        Some(ResultValues {
            passed,
            supply,
            consumption,
            impedance,
            known,
            unknown,
        })
    }
}

fn resolve_network<'a>(model: &'a Server, target: &'a str) -> Option<(&'a str, u8)> {
    if target.starts_with("//") {
        let parts = target
            .trim_start_matches('/')
            .split('/')
            .collect::<Vec<_>>();
        let [project, network] = parts.as_slice() else {
            return None;
        };
        return Some((project, network.parse().ok()?));
    }
    let network = target.parse().ok()?;
    Some((model.current.as_deref()?, network))
}

pub(crate) fn command(model: &Server, tag: &str, words: &[&str]) -> Response {
    if words.len() < 3 {
        return err(tag, 400, "400 Syntax Error.");
    }
    let Some(directory) = model.unitspec_dir.as_deref() else {
        return err(
            tag,
            408,
            "408 Operation failed: Calculation failed: No unit catalog available",
        );
    };
    let catalog = match load_catalog(directory) {
        Ok(catalog) => catalog,
        Err(_) => {
            return err(
                tag,
                408,
                "408 Operation failed: Calculation failed: No unit catalog available",
            )
        }
    };
    let Some((project_name, network_address)) = resolve_network(model, words[2]) else {
        return err(tag, 400, "400 Syntax Error.");
    };
    if model.current.as_deref() != Some(project_name) {
        return err(tag, 404, "404 Project not selected");
    }
    let Some(project) = model.projects.get(project_name) else {
        return err(
            tag,
            408,
            "408 Operation failed: Calculation failed: Can't find project",
        );
    };
    let Some(network) = project.networks.get(&network_address) else {
        return err(
            tag,
            408,
            "408 Operation failed: Calculation failed: Can't find network",
        );
    };
    let mut units = network.units.values().collect::<Vec<_>>();
    units.sort_by_key(|unit| unit.address);
    let Some(values) = catalog.calculate(units) else {
        return err(tag, 500, "500 Internal error.");
    };
    Response {
        tag: tag.to_string(),
        lines: vec![
            format!(
                "134-result: {}",
                if values.passed { "OK" } else { "FAILED" }
            ),
            format!("134-current_supply(mA)={}", values.supply),
            format!("134-current_consumption(mA)={}", values.consumption),
            format!("134-impedance(ohms)={:.1}", values.impedance as f64),
            format!("134-units_calculated={}", values.known),
        ],
        final_text: format!("134 units_not_calculated={}", values.unknown),
        status: 134,
    }
}
